src/codecs/bmp/decoder.rs

```diff
@@ -1256,226 +1133,207 @@ impl<R: Read + Seek> BmpDecoder<R> {
+                                    let xdelta = self.reader.read_u8()?;
+                                    let ydelta = self.reader.read_u8()?;
+                                    RLEInsn::Delta(xdelta, ydelta)
+                                }
+                                _ => {
+                                    let mut length = op as usize;
+                                    if self.image_type == ImageType::RLE4 {
+                                        length = (length + 1) / 2;
+                                    }
+                                    length += length & 1;
+                                    let mut buffer = vec![0; length];
+                                    self.reader.read_exact(&mut buffer)?;
+                                    RLEInsn::Absolute(op, buffer)
+                                }
+                            }
+                        }
+                        _ => {
+                            let palette_index = self.reader.read_u8()?;
+                            RLEInsn::PixelRun(control_byte, palette_index)
+                        }
+                    }
+                };
 
-        let mut delta_rows_left = 0;
-        let mut delta_pixels_left = skip_pixels;
-        let mut eof_hit = false;
-
-        // Scope the borrowing of pixel_data by the row iterator.
-        {
-            // Handling deltas in the RLE scheme means that we need to manually
-            // iterate through rows and pixels.  Even if we didn't have to handle
-            // deltas, we have to ensure that a single runlength doesn't straddle
-            // two rows.
-            let mut row_iter = self.rows(&mut pixel_data);
-            // If we have previously hit a delta value,
-            // blank the rows that are to be skipped.
-            blank_bytes((&mut row_iter).take(skip_rows.into()));
-            let mut insns_iter = RLEInsnIterator {
-                r: &mut self.reader,
-                image_type,
-            };
-            let p = self.palette.as_ref().unwrap();
-
-            'row_loop: while let Some(row) = row_iter.next() {
-                let mut pixel_iter = row.chunks_mut(num_channels);
-                // Blank delta skipped pixels if any.
-                blank_bytes((&mut pixel_iter).take(delta_pixels_left.into()));
-                delta_pixels_left = 0;
-
-                'rle_loop: loop {
-                    if let Some(insn) = insns_iter.next() {
-                        match insn {
-                            RLEInsn::EndOfFile => {
-                                blank_bytes(pixel_iter);
-                                blank_bytes(row_iter);
-                                eof_hit = true;
-                                break 'row_loop;
+                match instruction {
+                    RLEInsn::EndOfFile => {
+                        pixel_iter.for_each(|p| p.fill(0));
+                        row_iter.for_each(|r| r.fill(0));
```

Comment:
It is somewhat questionable whether we should return an error here if we're not already at the end of the row and at the last row of the image. The previous code seemed to be somewhat inconsistent about it (the logic around "eof_hit")